AG-13008, AG-13041 Fix Delete/Escape key deleting the annotation in text-edit mode

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-13008
https://ag-grid.atlassian.net/browse/AG-13041

olegat committed:
AG-13008 Bugfix: Delete/Backspace deletes the annotation when editing text

olegat committed:
AG-13008 Add e2e test for deleting text annotations

lsjroberts commented on packages/ag-charts-enterprise/src/features/annotations/annotationsStateMachine.ts:
```diff
@@ -390,8 +390,12 @@ export class AnnotationsStateMachine extends StateMachine<States, AnnotationType
 
                 delete: () => {
                     if (this.active == null) return;
+                    this.transition('cancel');
```
This is an anti-pattern, we should not transition inside these events. Partly as that could easily create an infinite loop, but also as it goes against the design of state machines.

olegat replied:
Fair. Another solution could be to store the `visible: boolean` state in the `TextInput` class and check that instead. I'll do that.

olegat replied:
Done.

lsjroberts reviewed (changes requested):
There's been a lot of back and forth on deleting annotations due to changes in interactions.

Please make sure any changes continue to work for both of these fixes:

- https://github.com/ag-grid/ag-charts/pull/2160
- https://github.com/ag-grid/ag-charts/pull/2454

lsjroberts commented on packages/ag-charts-website/e2e/toolbar.spec.ts:
```diff
@@ -126,4 +126,46 @@ test.describe('toolbar', () => {
         await expect(page).toHaveScreenshot('callout-6-change-fill-color.png', { animations: 'disabled' });
     });
+
+    test('AG-13008 delete annotation', async ({ page }) => {
```
Also, can we just put some simple tests around this into the existing `test('text', ...` block? These tests are supposed to be just smoke tests.

olegat replied:
I've added tests to `test('text', ...)` for the <kbd>Backspace</kbd> and <kbd>Delete</kbd> keys

olegat committed:
AG-13008 Fix SELECTORS.annotationOptionsDeleteButton

olegat committed:
Revert "AG-13008 Bugfix: Delete/Backspace deletes the annotation when editing text"

This reverts commit 5da45e6825c75cbb53182193d2dfa881363fcb12.

olegat committed:
AG-13008 Ignore KeyNavEvent<'delete'> when text input is shown

olegat committed:
AG-13008 CR Add simple tests to existing `test('text', ...)`

lsjroberts reviewed (approved):
LGTM, thanks!

olegat committed:
AG-13041 Don't show focus after deleting annotations with keyboard

This fixes a regression caused by the removal of the "mouseBlur" feature in
KeyNavManager.

This is sort of an anti-pattern. Showing the focus after pressing Backspace or
Delete on the keyboard is the "correct" behaviour. However, the correct
behaviour in this scenario feels flimsy.

Rather than restoring the "mouseBlur" feature, we instead create and focus on a
temporary so-called "focus mask" which is added above the canvas. The focus mask
does not respond to pointer events, and puts the canvas-proxy back into focus
whenever any key is pressed. The focus mask is removed once blurred. This mimics
the "mouseBlur" functionality that the use to have.

alantreadway reviewed (approved):
LGTM

olegat commented:
> There's been a lot of back and forth on deleting annotations due to changes in interactions.
> 
> Please make sure any changes continue to work for both of these fixes:
> 
> * [Ag 12476/annotations floating toolbar delete functionality #2160](https://github.com/ag-grid/ag-charts/pull/2160)
> * [AG-12598 Fix escape to cancel creating annotation #2454](https://github.com/ag-grid/ag-charts/pull/2454)

I've checked these.

✅ AG-12476 - No regression
✅ AG-12598 - No regression

olegat committed:
Revert "AG-13041 Don't show focus after deleting annotations with keyboard"

This reverts commit c3477c16533089eea573117c11ad0ff73a8fc959.

olegat committed:
AG-13037 Track second-last input event to avoid showing focus box when deleting annotation

olegat committed:
AG-13008 e2e: Click the page rather that canvas HTML element

olegat committed:
AG-13008 Fix e2e test: Must type text to create annotation

olegat commented:
Rebased and resolved conflict by dropping the e2e snapshot update to this file:
[packages/ag-charts-website/e2e/zoom.spec.ts-snapshots/zoom-crosshairs-after-wheel-zoom-chromium-linux.png](https://github.com/ag-grid/ag-charts/blob/28b713bb53099125953b647ef0827e89c9a39897/packages/ag-charts-website/e2e/zoom.spec.ts-snapshots/zoom-crosshairs-after-wheel-zoom-chromium-linux.png)

olegat committed:
AG-13008 Update e2e test & add snapshot
